### PR TITLE
This change implements etag-addressing for our cache.

### PR DESCRIPTION
**Pull request by mattmoor**
ref: https://github.com/chainguard-dev/go-apk/issues/63

### CODE REVIEW

File: pkg/apk/repo_test.go [hunk 1]
```diff
@@ -142,66 +166,91 @@ func TestGetRepositoryIndexes(t *testing.T) {
...
+
 		// check that the contents are the same
-		index1, err := os.ReadFile(cacheIndexFile)
-		require.NoError(t, err, "unable to read cache index file")
-		index2, err := os.ReadFile(filepath.Join(testAlternatePkgDir, indexFilename))
-		require.NoError(t, err, "unable to read previous index file")
-		require.Equal(t, index1, index2, "index files do not match")
+		require.NotEqual(t, index1, index2, "index files do not match")
+	})
+	t.Run("test cache concurrency", func(t *testing.T) {
```
Comment (mattmoor) [inline pkg/apk/repo_test.go:231]: This test fails immediately (over and over, as we have been seeing) at HEAD.

With this change it passes.

**Review by jonjohnsonjr — APPROVED** [review]

**Review by imjasonh — APPROVED** [review]
Nice.

I'd love to have this depend on the digest of the contents instead of relying on the server's etag, but we can make that change later if we're worried.

**Comment by mattmoor** [comment]
It looks like `go test -race` cannot handle my unit test, I'm going to dial down the thread count

**Comment by mattmoor** [comment]
FWIW locally:
```
go test ./... -race
?       github.com/chainguard-dev/go-apk/pkg/logger     [no test files]
?       github.com/chainguard-dev/go-apk/pkg/signature  [no test files]
ok      github.com/chainguard-dev/go-apk/pkg/apk        42.451s
ok      github.com/chainguard-dev/go-apk/pkg/fs 0.333s
ok      github.com/chainguard-dev/go-apk/pkg/passwd     0.514s
ok      github.com/chainguard-dev/go-apk/pkg/tarball    0.390s
```

**Pull request merged**